modules/tinymce/src/plugins/lists/main/ts/actions/ToggleList.ts

```diff
@@ -76,9 +84,12 @@ const getEndPointNode = (editor: Editor, rng: Range, start: Boolean, root: Node)
   const findBetterContainer = (container: Node, forward: boolean): Optional<Node> => {
     const walker = new DomTreeWalker(container, root);
     const dir = forward ? 'next' : 'prev';
+    const blockParent = SugarElement.fromDom(findBlockAncestor(container));
     let node;
     while ((node = walker[dir]())) {
-      if (!(NodeType.isVoid(editor, node) || Unicode.isZwsp(node.textContent as string) || node.textContent?.length === 0)) {
+      const suitableType = !(NodeType.isVoid(editor, node) || Unicode.isZwsp(node.textContent as string) || node.textContent?.length === 0);
+      const isWithinParent = Has.ancestor(SugarElement.fromDom(node), blockParent);
```

Comment:
Since this seems to just make sure that the node we find isn't outside the scope of the parent block then you could just feed the DomTreeWalker the parent node as a scope if it exists otherwise fallback to the root. So this function then becomes:

```
  const findBetterContainer = (container: Node, forward: boolean): Optional<Node> => {
    const scope = editor.dom.getParent(container, editor.dom.isBlock) ?? root;
    const walker = new DomTreeWalker(container, scope);
    const dir = forward ? 'next' : 'prev';
    let node;
    while ((node = walker[dir]())) {
      if (!(NodeType.isVoid(editor, node) || Unicode.isZwsp(node.textContent as string) || node.textContent?.length === 0)) {
        return Optional.some(node);
      }
    }

    return Optional.none();
  };
```

Where the actual two lines changes are:
```
    const scope = editor.dom.getParent(container, editor.dom.isBlock) ?? root;
    const walker = new DomTreeWalker(container, scope);
```

That seems to the same thing an the tests seems to pass at least the `ApplyTest.ts`.

Reply:
I'd still say we should step up until we find the block, just to prevent a cases where inlines are stacked. But yes, this is a much neater solution which I should have thought of.

Reply:
Right I guess it should be `const scope = editor.dom.getParent(container, editor.dom.isBlock) ?? container;` since the function you have either finds a block parent or falls back to the container. But I guess we could keep the function to if that is clearer.